### ISSUE TITLE
Consider not enabling replace refs by default

### ISSUE THREAD

**Issue by bk2204** [newren/git-filter-repo#235]
The documentation mentions that `git filter-repo` produces replace refs by default, “which if pushed and fetched will allow users to continue to refer to new commits using (unabbreviated) old commit IDs.”  In theory, I agree with this, but the practice is different.

First of all, Git neither fetches nor pushes replace refs by default, so unless the user is intimately aware of how replace refs work, this functionality doesn't work at all since they're never transferred.  It's also difficult to push only the replace refs relating to the work you're pushing without also pushing all of the ones for unrelated, in-progress work.  Moreover, replace refs don't work with libgit2, so anything that uses libgit2, like a Rust-based project, or GitHub (and, I believe GitLab), doesn't support them.  Since two of the largest Git hosting providers don't support replace refs, even when used properly. they're not especially useful for many users.

In addition, I got a user report where the user thought they had rewritten their branch with `git filter-repo`, but had pushed something different to the remote.  They were then confused why what they saw in their local repository, which honored replace refs, didn't match what they saw in the web interface, which didn't.  They were understandably confused and angry when the problem was explained to them because they expected things to just work.

Since replace refs are one of those esoteric features that sound cool in practice but tend to have a bunch of pain points that lead to user confusion, I'd like to suggest that they not be enabled by default.

**Comment by micimize** [comment]
This would be quite a fundamental change – I think a designated guide on recommendations for working with GitHub, etc and situations where pushing to a fresh repo isn't an option is maybe more reasonable starting point.

The other side of this coin is, shouldn't [the git ecosystem](https://github.com/libgit2/libgit2/issues/1730) support [a feature that's been around since 2.1.4](https://git-scm.com/docs/git-replace/2.1.4)?

**Comment by newren** [comment]
Hi Brian,

Thanks for the report about user actions and about potential ideas for improving the documentation.

In addition to the issues you point out, there are also the facts that many refs might slow down git operations, that commit-graphs are turned off by the presence of replace refs (instead of just turning off when there is a replace ref for an existing commit object), and the recent bugs I reported about replace refs (https://lore.kernel.org/git/CABPp-BEAbN05+hCtK=xhGg5uZFqbUvH9hMcCNMcBWp5JWLqzPw@mail.gmail.com/).  You can also add jgit and Gerrit to the list of services that don't support replace refs (at least not the versions I tried).

Some comments on your individual points...

> The documentation mentions that `git filter-repo` produces replace refs by default, “which if pushed and fetched will allow users to continue to refer to new commits using (unabbreviated) old commit IDs.” In theory, I agree with this, but the practice is different.
> 
> First of all, Git neither fetches nor pushes replace refs by default.

Right, that's why I said "if pushed and fetched", otherwise I would have just left that phrase out.  Since it's still obviously misinterpreted, though, I guess I can insert a "manually" in there to drive the point home.

> It's also difficult to push only the replace refs relating to the work you're pushing without also pushing all of the ones for unrelated, in-progress work.

I don't find this very relevant: This is already a really unlikely situation, since as you note replace refs are "one of those esoteric features", meaning folks are unlikely to have used them.  Even if they do have them, wouldn't you want them included?  fast-export automatically will heed and use them as part of the rewrite and they are updated by default (if you have a replace ref from A->B and B is rewritten to C, filter-repo changes the replace ref to be A->C), and there are lots of options about how to delete/update/append to the replace refs.  Surely if you're rewriting history, you should be aware of what exists in the history and how you are rewriting.  And if you really want an option for those to just be left out of the rewrite and don't want to push them, then you can just reclone?  The replace refs aren't grabbed by default with fetch/clone after all.  On top of all of that, the filter-repo documentation pushes very strongly for people to make a clean clone before doing history rewriting and there are several checks in place to try to reinforce that.

> Moreover, replace refs don't work with libgit2, so anything that uses libgit2, like a Rust-based project, or GitHub (and, I believe GitLab), doesn't support them. Since two of the largest Git hosting providers don't support replace refs, even when used properly. they're not especially useful for many users.

I agree with your first sentence; that's an important point.  As I noted above, you can also add jgit and Gerrit to the mix.  In fact, I explicitly documented that, "Sadly, some existing git servers (e.g. Gerrit, GitHub) do not yet understand replace refs, and thus one can’t use old commit hashes within their UI; this may change in the future."  I should probably add libgit2 and jgit to the list of things that don't support replace refs.  But your second statement does not follow from the first; as also documented, "But replace refs at least help users locally within the git CLI."

> In addition, I got a user report where the user thought they had rewritten their branch with `git filter-repo`, but had pushed something different to the remote. They were then confused why what they saw in their local repository, which honored replace refs, didn't match what they saw in the web interface, which didn't. They were understandably confused and angry when the problem was explained to them because they expected things to just work.

Interesting.  git-filter-repo documentation starts with:

"Rapidly rewrite entire repository history using user-specified filters. This is a destructive operation which should not be used lightly; it writes new commits, trees, tags, and blobs corresponding to (but filtered from) the original objects in the repository, then deletes the original history and leaves only the new. See [DISCUSSION] for more details on the ramifications of using this tool."

The DISCUSSION section includes some *very* strongly worded warnings about not pushing back to the same repo and why doing so is fraught with errors:

"""
Using filter-repo is relatively simple, but rewriting history is part of a larger discussion in terms of collaboration. When you rewrite history, the old and new histories are no longer compatible; if you push this history somewhere for others to view, it will look as though you’ve done a rebase of all branches and tags. Make sure you are familiar with the "RECOVERING FROM UPSTREAM REBASE" section of git-rebase(1) (and in particular, "The hard case") before proceeding, in addition to this section.

Steps to use git-filter-repo as part of the bigger picture of doing a history rewrite are roughly as follows:

...

4. Push your new repository to its new home (note that refs/remotes/origin/* will have been moved to refs/heads/* as the first part of filter-repo, so you can just deal with normal branches instead of remote tracking branches). While you can force push this to the same URL you cloned from, there are good reasons to consider pushing to a different location instead:

* People who cloned from the original repo will have old history. When they fetch the new history you force pushed up, unless they do a git reset --hard @{u} on their branches or rebase their local work, git will think they have hundreds or thousands of commits with very similar commit messages as what exist upstream (but which include files you wanted excised from history), and allow the user to merge the two histories, resulting in what looks like two copies of each commit. If they then push this history back up, then everyone now has history with two copies of each commit and the bad files have returned. You’re more likely to succeed in forcing people to get rid of the old history if they have to clone a new URL.

* Rewriting history will rewrite tags; those who have already downloaded tags will not get the updated tags by default (see the "On Re-tagging" section of git-tag(1)). Every user trying to use an existing clone will have to forcibly delete all tags and re-fetch them; it may be easier for them to just re-clone, which they are more likely to do with a new clone URL.

* Rewriting history may delete some refs (e.g. branches that only had files that you wanted excised from history); unless you run git push with the --mirror or --prune options, those refs will continue to exist on the server. If folks then merge these branches into others, then people have started mixing old and new history. If users had already cloned these branches, removing them from the server isn’t enough; you need all users to delete any local branches based on these refs and run fetch with the --prune option as well. Simply re-cloning from a new URL is easier.

* The server may not allow you to force push over some refs. For example, code review systems may have special ref namespaces (e.g. refs/changes/, refs/pull/, refs/merge-requests/) that they have locked down.

5. If you still want to push your rewritten history back to the original url despite my warnings above, you’ll have to manage it very carefully:

git-filter-repo deletes the "origin" remote to help avoid people accidentally repushing to the same repository, so you’ll need to remind git what origin’s url was. You’ll have to look up the command for that.

* You’ll need to carefully synchronize with everyone who has cloned the repository, and will also need to carefully synchronize with everything (e.g. CI systems) that has cloned it. Every single clone will either need to be thrown away and re-cloned, or need to take all the steps outlined in item 4 as well as follow the necessary steps from "RECOVERING FROM UPSTREAM REBASE" section of git-rebase(1). If you miss fixing any clones, you’ll risk mixing old and new history and end up with an even worse mess to clean up.

* Finally, you’ll need to consult any documentation from your hosting provider about how to remove any server-side references to the old commits (example: GitLab’s docs on reducing repository size).
"""

It also ends with a note about how replace refs require manual effort to push and pull and notes how existing repository managers do not support them:

"""
(Optional) filter-repo by default creates replace refs (see git-replace(1)) for each rewritten commit ID, allowing you to use old (unabbreviated) commit hashes to refer to the newly rewritten commits. If you want to use these replace refs, push them to the relevant clone URL and tell users to adjust their fetch refspec (e.g. git config --add remote.origin.fetch +refs/replace/*:refs/replace/*) Sadly, some existing git servers (e.g. Gerrit, GitHub) do not yet understand replace refs, and thus one can’t use old commit hashes within their UI; this may change in the future. But replace refs at least help users locally within the git CLI.
"""

> Since replace refs are one of those esoteric features that sound cool in practice but tend to have a bunch of pain points that lead to user confusion, I'd like to suggest that they not be enabled by default.

I like to avoid pain points for users, but in this case, history rewriting is inherently a dangerous operation.  Everything about it is unsafe, particularly if folks decide to ignore the documentation.  And replace refs are extremely useful in that they provide a short term way for folks to continue referring to commits at the command line using old commit IDs, helping them during the transition period after a rewrite (and then at some point they can likely discard the replace refs after that transition period).

**Comment by newren** [comment]
I pushed commit 4e704d2ef1 (git-filter-repo.txt: add some clarifications around replace refs, 2021-04-07) to: insert "manually", mention libgit2 and jgit and GitLab, and mention the commit-graph behavior.  If there are other places that should be touched up too or other things I should mention, let me know.

Thanks for highlighting this!

**Comment by newren** [comment]
I also just pushed commit 75e67bcd44 (git-filter-repo.txt: link to GitHub docs on purging old history, 2021-04-07), to link to GitHub docs on purging old history and the nice caveats it mentions there about how cached views and forks and clones aren't updated by the process.  However, note that my link is of the form:

> Finally, you’ll need to consult any documentation from your hosting provider about how to remove any server-side references to the old commits (example: [GitLab’s excellent docs on reducing repository size](https://docs.gitlab.com/ee/user/project/repository/reducing_the_repo_size_using_git.html), or just the warning box that references "GitHub support" from [GitHub’s otherwise dangerously out-of-date docs on removing sensitive data](https://docs.github.com/en/github/authenticating-to-github/removing-sensitive-data-from-a-repository)).

I'll be happy to remove the warning about the GitHub docs, once they are updated to stop recommending git-filter-branch [against git-filter-branch's own recommendation](https://git-scm.com/docs/git-filter-branch#_warning).

**Comment by bk2204** [comment]
> The other side of this coin is, shouldn't [the git ecosystem](https://github.com/libgit2/libgit2/issues/1730) support [a feature that's been around since 2.1.4](https://git-scm.com/docs/git-replace/2.1.4)?

libgit2, like many open source projects, is lacking interested parties to implement that feature.  I personally don't see any value from replace refs and haven't chosen to implement them.  If you think they provide value and want to implement them, please do.  I'm sure such a contribution would be very welcome.

> > Finally, you’ll need to consult any documentation from your hosting provider about how to remove any server-side references to the old commits (example: [GitLab’s excellent docs on reducing repository size](https://docs.gitlab.com/ee/user/project/repository/reducing_the_repo_size_using_git.html), or just the warning box that references "GitHub support" from [GitHub’s otherwise dangerously out-of-date docs on removing sensitive data](https://docs.github.com/en/github/authenticating-to-github/removing-sensitive-data-from-a-repository)).

I don't think, given the status quo, I would recommend that GitHub update their documentation to recommend git filter-repo at the moment, since generally the goal is to surgically remove a specific item and keep the rest of the history intact.  I don't think git filter-branch is entirely terrible, although I agree it is inefficient.  You are, of course, free to ask them to do something different if you disagree; the folks doing the documentation may feel git filter-repo is a better choice.

I should point out that my views here are my own and this issue was opened on my own behalf.  I do think replace refs are surprising and poorly supported and I don't recommend them to others, but it's clear you disagree, and it's your project to run as you see fit.

**Comment by newren** [comment]
Not sure what happened; it appears my previous response has disappeared.  So let's try again...

> I don't think, given the status quo, I would recommend that GitHub update their documentation to recommend git filter-repo at the moment, since generally the goal is to surgically remove a specific item and keep the rest of the history intact. I don't think git filter-branch is entirely terrible, although I agree it is inefficient. You are, of course, free to ask them to do something different if you disagree; the folks doing the documentation may feel git filter-repo is a better choice.

Recommending git filter-repo might be a nice bonus, but the important bit is to stop recommending git filter-branch.  In my estimation git filter-branch is an entirely terrible tool, and I'd argue that filter-branch itself agrees with me:

```
$ git filter-branch -h
WARNING: git-filter-branch has a glut of gotchas generating mangled history
	 rewrites.  Hit Ctrl-C before proceeding to abort, then use an
	 alternative filtering tool such as 'git filter-repo'
	 (https://github.com/newren/git-filter-repo/) instead.  See the
	 filter-branch manual page for more details; to squelch this warning,
	 set FILTER_BRANCH_SQUELCH_WARNING=1.
```

It appears this issue has already been reported (github/docs#3617).  If others are intent on recommending dangerous tools, that is their choice, but I cannot in good conscience link to such pages without adding a disclaimer.
 
> I should point out that my views here are my own and this issue was opened on my own behalf. I do think replace refs are surprising and poorly supported and I don't recommend them to others, but it's clear you disagree, and it's your project to run as you see fit.

In most cases, I'd probably avoid replace refs as well given all the caveats (lack of support in other implementations like libgit2 and jgit; lack of support in hosting sites like Gerrit, GitHub, and GitLab; known bugs in the implementation in git.git), and the fact that more than one of my fellow git developers (including you) frown on them.  But, in this case, they solve a very specific short term usecase so nicely that I think they are very much warranted.

**Comment by newren** [comment]
I'm reopening this.  @bk2204 certainly made excellent points (he always does), but at the time I didn't think on the whole it merited changing the default.  A few things have happened since then:

  * A few more reports from users misunderstanding/misusing replace refs, in new and interesting ways
  * Reflecting on https://lore.kernel.org/git/CABPp-BEAbN05+hCtK=xhGg5uZFqbUvH9hMcCNMcBWp5JWLqzPw@mail.gmail.com/ has made me a bit more uneasy about foot-guns within git itself (not even worrying about libgit2, jgit, or the code hosting sites) for users
  *  (The biggest one, for me) There was a history rewrite that we had wanted to do for years and years at ${FORMER_EMPLOYER} and for which at various meetings I had been told many, many times over the years that we still needed to be able to look up things by *old* commit IDs or the conversion was a no-go.  This rewrite finally happened at the very end of 2021 (shrunk the repo by a factor greater than 4, providing very nice improvements to clone times).  Due to the commit-graph hammer that disables itself if there are any replace refs, these were pushed to a separate repository so that only those who needed them could get them.  And then I watched as there was almost no use of the old commit IDs, and any use there was was limited to the first week.  So much for years of being repeatedly hounded about the importance of being able to look things up by old commit IDs...

So, the cons for replace-refs-by-default were bigger than I appreciated, and the pros for it were much smaller than I suspected.  However, I still think there are folks who would be blocked on doing some good necessary cleanups without the replace-refs feature I have in git-filter-repo, so I'm glad we have it.  But I think it's time to change the default, as brian suggests.  (I've thought this for about two years, but I "went missing" on git-filter-repo stuff so just kept it in the back of my mind, and then accumulated lots of pull requests that I needed to respond to first...)
